torch/nn/quantizable/modules/activation.py

```diff
@@ -0,0 +1,439 @@
+        information
+
+    Allows the model to jointly attend to information from different
+    representation subspaces.
+    See reference: Attention Is All You Need
+
+    The original MHA module is not quantizable.
+    This reimplements it by explicitly instantiating the linear layers.
+
+    .. math::
+        \text{MultiHead}(Q, K, V) = \text{Concat}(head_1,\dots,head_h)W^O
+        \text{where} head_i = \text{Attention}(QW_i^Q, KW_i^K, VW_i^V)
+
+    Args:
+        embed_dim: total dimension of the model.
+        num_heads: parallel attention heads.
+        dropout: a Dropout layer on attn_output_weights. Default: 0.0.
+        bias: add bias as module parameter. Default: True.
+        add_bias_kv: add bias to the key and value sequences at dim=0.
+        add_zero_attn: add a new batch of zeros to the key and
+                       value sequences at dim=1.
+        kdim: total number of features in key. Default: None.
+        vdim: total number of features in value. Default: None.
+
+        Note: if kdim and vdim are None, they will be set to embed_dim such that
+        query, key, and value have the same number of features.
+
+    Examples::
+
+        >>> import torch.nn.quantizable as nnqa
+        >>> multihead_attn = nnqa.MultiheadAttention(embed_dim, num_heads)
+        >>> attn_output, attn_output_weights = multihead_attn(query, key, value)
+
+    Note::
+        Please, follow the quantization flow to convert the quantizable MHA.
+    """
+    def __init__(self, embed_dim: int, num_heads: int,
+                 dropout: float = 0., bias: bool = True,
+                 add_bias_kv: bool = False, add_zero_attn: bool = False,
+                 kdim: int = None, vdim: int = None):
+        super(MultiheadAttention, self).__init__(embed_dim, num_heads, dropout,
+                                                 bias, add_bias_kv,
+                                                 add_zero_attn, kdim, vdim)
+        self.linear_Q = nn.Linear(self.embed_dim, self.embed_dim, bias=bias)
+        self.linear_K = nn.Linear(self.kdim, self.embed_dim, bias=bias)
+        self.linear_V = nn.Linear(self.vdim, self.embed_dim, bias=bias)
+
+        # TODO: The use of the `_LinearWithBias` increases the quantization noise
+        # The `out_proj` in the parent is ``_LinearWithBias`, so need to ignore
+        # the type for mypy not to complain.
+        self.out_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=True)  # type: ignore
+
+        # Functionals
+        self.q_scaling_product = nnq.FloatFunctional()
+
+        # Quant/Dequant
+        self.quant_attn_output = torch.quantization.QuantStub()
+        self.quant_attn_output_weights = torch.quantization.QuantStub()
+        self.dequant_q = torch.quantization.DeQuantStub()
+        self.dequant_k = torch.quantization.DeQuantStub()
+        self.dequant_v = torch.quantization.DeQuantStub()
+
+    def _get_name(self):
+        return 'QuantizableMultiheadAttention'
```

Comment:
is this going to return `QuantizableMultiheadAttention` in both fp32 with observers and quantized state?  If yes, would it be possible to refactor the code so that the final quantized module prints out `QuantizedMultiheadAttention`, to help with debugging?

Reply:
It is not as straight-forward. This will require reflection design pattern with the change of function when the `from_observed` is called. I would not recommend doing it, as this is a power-feature. Alternatively, one can check if the layer weights are quantized, and change the name accordingly. But I think it might be error prone -- I'll think about it...

Reply:
it would be good to fix this in the long term, the current state will be confusing to people who use this without reading the code